File: Server/src/tools/validation.py
"""Validation tools for Fusion 360 MCP Server.

Contains functions for testing connections, querying model state, and utility operations.
Uses @fusion_tool decorator for automatic HTTP handling and telemetry.
"""

from .base import fusion_tool


@fusion_tool
def test_connection():
    """Test the connection to the Fusion 360 Add-In server."""


@fusion_tool(method="GET")
def get_model_state():
    """
    Get the current state of the Fusion 360 model.

    Returns:
        Dictionary with:
        - body_count: Number of bodies in the model
        - sketch_count: Number of sketches
        - bodies: List with details for each body (name, volume, bounding box)
        - sketches: List with details for each sketch (name, profile count)
        - design_name: Name of the active design
    """


@fusion_tool(method="GET")
def get_faces_info(body_index: int = 0):
    """
    Get detailed face information for a body.

    Args:
        body_index: Index of the body to inspect

    Returns:
        Dictionary with:
        - face_count: Number of faces
        - faces: List with details for each face (index, type, area, centroid)
    """


@fusion_tool
def delete_all():
    """Delete all objects in the current Fusion 360 session."""


@fusion_tool
def undo():
    """Undo the last action."""

File: Server-TS/src/tools/validation.ts
/**
 * Validation/Infrastructure tools for Fusion 360 MCP Server.
 */

import { z } from "zod";
import { createGetTool, createPostTool, ToolResult } from "./base.js";

// Tool implementations using factory functions
const _testConnection = createPostTool("test_connection", "test_connection", true);
const _getModelState = createGetTool("get_model_state");
const _undo = createPostTool("undo", "undo", true);
const _deleteAll = createPostTool("delete_everything", "delete_everything", true);
const _getFacesInfo = createGetTool("get_faces_info");

// Zod schemas for validation
export const testConnectionSchema = z.object({});

export const getModelStateSchema = z.object({});

export const undoSchema = z.object({});

export const deleteAllSchema = z.object({
  bodies: z.boolean().default(true).describe("Delete all bodies"),
  sketches: z.boolean().default(true).describe("Delete all sketches"),
  construction: z
    .boolean()
    .default(true)
    .describe("Delete construction geometry"),
  parameters: z.boolean().default(false).describe("Delete user parameters"),
});

export const getFacesInfoSchema = z.object({
  body_index: z.number().default(0).describe("Index of the body to inspect"),
});

// Exported tool functions

/**
 * Test the connection to the Fusion 360 Add-In.
 */
export async function test_connection(): Promise<ToolResult> {
  return _testConnection();
}

/**
 * Get the current state of the model including body count, sketch count, etc.
 */
export async function get_model_state(): Promise<ToolResult> {
  return _getModelState();
}

/**
 * Undo the last action in Fusion 360.
 */
export async function undo(): Promise<ToolResult> {
  return _undo();
}

/**
 * Delete all objects in the current Fusion 360 session.
 */
export async function delete_all(
  bodies: boolean = true,
  sketches: boolean = true,
  construction: boolean = true,
  parameters: boolean = false
): Promise<ToolResult> {
  return _deleteAll({ bodies, sketches, construction, parameters });
}

/**
 * Get detailed face information for a body.
 */
export async function get_faces_info(
  body_index: number = 0
): Promise<ToolResult> {
  return _getFacesInfo({ body_index });
}
